velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -126,61 +126,62 @@ struct MinMaxByNAccumulator {
 
   void
   compareAndAdd(C comparison, std::optional<V> value, Compare& comparator) {
-    if (topPairs.size() < n) {
-      topPairs.push({comparison, value});
+    if (heapValues.size() < n) {
+      heapValues.push_back({comparison, value});
+      std::push_heap(heapValues.begin(), heapValues.end(), comparator);
     } else {
-      const auto& topPair = topPairs.top();
+      const auto& topPair = heapValues.front();
       if (comparator.compare(comparison, topPair)) {
-        topPairs.pop();
-        topPairs.push({comparison, value});
+        std::pop_heap(heapValues.begin(), heapValues.end(), comparator);
+        heapValues.back() = std::make_pair(comparison, value);
+        std::push_heap(heapValues.begin(), heapValues.end(), comparator);
       }
     }
   }
 
-  /// Moves all values from 'topPairs' into 'rawValues' and 'rawValueNulls'
-  /// buffers. The queue of 'topPairs' will be empty after this call.
+  /// Extract all values from 'heapValues' into 'rawValues' and 'rawValueNulls'
+  /// buffers. The heap remains unchanged after the call.
   void extractValues(
       TRawValue* rawValues,
       uint64_t* rawValueNulls,
-      vector_size_t offset) {
-    const vector_size_t size = topPairs.size();
-    for (auto i = size - 1; i >= 0; --i) {
-      const auto& topPair = topPairs.top();
+      vector_size_t offset,
+      Compare& comparator) {
+    std::sort_heap(heapValues.begin(), heapValues.end(), comparator);
+    for (int64_t i = heapValues.size() - 1; i >= 0; --i) {
+      const auto& pair = heapValues[i];
       const auto index = offset + i;
-
-      const bool valueIsNull = !topPair.second.has_value();
+      const bool valueIsNull = !pair.second.has_value();
       bits::setNull(rawValueNulls, index, valueIsNull);
       if (!valueIsNull) {
-        RawValueExtractor<V>::extract(rawValues, index, topPair.second.value());
+        RawValueExtractor<V>::extract(rawValues, index, pair.second.value());
       }
-
-      topPairs.pop();
     }
+    std::make_heap(heapValues.begin(), heapValues.end(), comparator);
   }
 
-  /// Moves all pairs of (comparison, value) from 'topPairs' into
-  /// 'rawComparisons', 'rawValues' and 'rawValueNulls' buffers. The queue of
-  /// 'topPairs' will be empty after this call.
+  /// Moves all pairs of (comparison, value) from 'heapValues' into
+  /// 'rawComparisons', 'rawValues' and 'rawValueNulls' buffers. The heap
+  /// remains unchanged after the call.
   void extractPairs(
       TRawComparison* rawComparisons,
       TRawValue* rawValues,
       uint64_t* rawValueNulls,
-      vector_size_t offset) {
-    const vector_size_t size = topPairs.size();
-    for (auto i = size - 1; i >= 0; --i) {
-      const auto& topPair = topPairs.top();
+      vector_size_t offset,
+      Compare& comparator) {
+    std::sort_heap(heapValues.begin(), heapValues.end(), comparator);
+    for (int64_t i = heapValues.size() - 1; i >= 0; --i) {
```

Comment:
nit: might be worth mentioning with a single line comment that we are returning in ascending order since it looks like we have maintained that expectation here as compared to the previous implementation using PQ. Also, is this a required expectation on the order here?

Reply:
@kagamiori I think i mistakingly wrote ascending order instead of descending in my comment. Your previous change was correct and my comment was just asking you to write the order in a code comment. The original implementation used PQ and pulling one value at a time from there would return the results in descending order (which is what you were doing earlier by iterating from size()-1 to 0 index of the array).

Secondly, Can you please check if there indeed is any strict order that is expected here?

Reply:
@kagamiori Thanks for the detailed reply, looks like I got confused in my second review instead haha😅